packages/apollo-engine-reporting/src/durationHistogram.ts

```diff
@@ -0,0 +1,57 @@
+export class DurationHistogram {
+  private readonly buckets: number[];
+  public static readonly BUCKET_COUNT = 384;
+  public static readonly EXPONENT_LOG = Math.log(1.1);
+
+  public toArray(): number[] {
+    let bufferedZeroes = 0;
+    const outputArray: number[] = [];
+
+    for (const value of this.buckets) {
+      if (value === 0) {
+        bufferedZeroes++;
+      } else {
+        if (bufferedZeroes === 1) {
+          outputArray.push(0);
+        } else if (bufferedZeroes !== 0) {
+          outputArray.push(-bufferedZeroes);
+        }
+        outputArray.push(value);
+        bufferedZeroes = 0;
+      }
+    }
+    return outputArray;
+  }
+
+  static durationToBucket(durationNs: number): number {
+    const log = Math.log(durationNs / 1000.0);
+    const unboundedBucket = Math.ceil(log / DurationHistogram.EXPONENT_LOG);
+
+    // Comapre <= 0 to catch -0
+    return (unboundedBucket <= 0 || Number.isNaN(unboundedBucket)) ?
+      0 :
+      (unboundedBucket >= DurationHistogram.BUCKET_COUNT) ? DurationHistogram.BUCKET_COUNT - 1 : unboundedBucket;
+  }
+
+
+  public incrementDuration(durationNs: number) {
+    this.incrementBucket(DurationHistogram.durationToBucket(durationNs));
+  }
+
+  public incrementBucket(bucket: number) {
+    this.buckets[bucket]++;
+  }
+
+  public combine(otherHistogram: DurationHistogram) {
+    for (let i = 0; i < otherHistogram.buckets.length; i++) {
+      this.buckets[i] += otherHistogram.buckets[i];
+    }
+  }
+
+  constructor(buckets?: number[]) {
+    this.buckets = Array<number>(DurationHistogram.BUCKET_COUNT).fill(0);
```

Comment:
Most of the buckets are generally empty (eg, 1 minute is only bucket 188 so half of the buckets are only used for durations over a minute). It might be a better use of memory to grow the bucket array lazily, starting with the list either empty or some reasonable smaller size.

To grow the array from length L to M you can do `this.buckets.length = M; this.buckets.fill(0, L)` I think.  incrementBucket and combine would both have to ensure there are enough buckets.

Maybe this is a fancy thing we could do later? But I think it's straightforward enough and would save a fair amount of memory.

Reply:
I agree. I'm worried somewhat about perf, so I think this should be an adjustable value. I'm also wondering if the arrays get fairly sparse we could use a half-compacted format, but that would be a latter follow up.